src/Appwrite/Platform/Tasks/Backup.php

```diff
@@ -0,0 +1,318 @@
+        $filename = basename($file);
+
+        if (!$this->s3->exists('/')) {
+            throw new Exception('Can\'t read s3 root directory');
+        }
+
+        $destination = $this->s3->getRoot() . '/' . $filename;
+
+        try {
+            if (!$local->transfer($file, $destination, $this->s3)) {
+                throw new Exception('Error uploading to ' . $destination);
+            }
+        } catch (Exception $e) {
+            throw new Exception($e->getMessage());
+        }
+
+        if (!$this->s3->exists($destination)) {
+            throw new Exception('File not found in destination: ' . $destination);
+        }
+
+        if (!unlink($file)) {
+            throw new Exception('Error deleting: ' . $file);
+        }
+
+        self::log('Upload took ' . (microtime(true) - $start) . ' seconds');
+    }
+
+    public static function log(string $message): void
+    {
+        if (!empty($message)) {
+            Console::log(date('Y-m-d H:i:s') . ' ' . $message);
+        }
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function checkEnvVariables(): void
+    {
+        foreach (
+            [
+                '_APP_CONNECTIONS_BACKUPS_STORAGE',
+                '_APP_CONNECTIONS_DB_REPLICAS',
+            ] as $env
+        ) {
+            if (empty(App::getEnv($env))) {
+                throw new Exception('Can\'t read ' . $env);
+            }
+        }
+    }
+
+    public function getDsn(string $database): ?DSN
+    {
+        foreach (explode(',', App::getEnv('_APP_CONNECTIONS_DB_REPLICAS', '')) as $project) {
+            [$db, $dsn] = explode('=', $project);
+            if ($db === $database) {
+                return new DSN($dsn);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function setContainerId()
+    {
+        $stdout = '';
+        $stderr = '';
+        Console::execute('docker ps -f "name=xtrabackup" --format "{{.ID}}"', '', $stdout, $stderr);
+        if (!empty($stderr)) {
+            throw new Exception('Error setting container Id: ' . $stderr);
+        }
+
+        $containerId = str_replace(PHP_EOL, '', $stdout);
+        if (empty($containerId)) {
+            throw new Exception('Xtrabackup Container ID not found');
+        }
+
+        $this->xtrabackupContainerId = $containerId;
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function setProcessors()
+    {
+        $stdout = '';
+        $stderr = '';
+        Console::execute('docker exec ' . $this->xtrabackupContainerId . ' nproc', '', $stdout, $stderr);
+        if (!empty($stderr)) {
+            throw new Exception('Error setting processors: ' . $stderr);
+        }
+
+        $processors = str_replace(PHP_EOL, '', $stdout);
+        $processors = intval($processors);
+
+        if ($processors === 0) {
+            throw new Exception('Set Processors Error');
+        }
+
+        $this->processors = $processors;
```

Comment:
Let's leave some cores free

```suggestion
        $this->processors = \max(1, $processors - 2);
```